internal/resolver/delegatingresolver/delegatingresolver_ext_test.go

```diff
@@ -0,0 +1,540 @@
+// The test verifies that the delegating resolver combines unresolved proxy
+// host and target addresses correctly, returning addresses with the proxy host
+// populated and the target address included as an attribute.
+func (s) TestDelegatingResolverForMutipleProxyAddress(t *testing.T) {
+	const (
+		targetTestAddr          = "test.com"
+		resolvedTargetTestAddr1 = "1.1.1.1:8080"
+		resolvedTargetTestAddr2 = "2.2.2.2:8080"
+		envProxyAddr            = "proxytest.com"
+		resolvedProxyTestAddr1  = "11.11.11.11:7687"
+		resolvedProxyTestAddr2  = "22.22.22.22:7687"
+	)
+	hpfe := func(req *http.Request) (*url.URL, error) {
+		if req.URL.Host == targetTestAddr {
+			return &url.URL{
+				Scheme: "https",
+				Host:   envProxyAddr,
+			}, nil
+		}
+		t.Errorf("Unexpected request host to proxy: %s want %s", req.URL.Host, targetTestAddr)
+		return nil, nil
+	}
+	originalhpfe := internal.HTTPSProxyFromEnvironment
+	internal.HTTPSProxyFromEnvironment = hpfe
+	defer func() {
+		internal.HTTPSProxyFromEnvironment = originalhpfe
+	}()
+
+	targetResolver := manual.NewBuilderWithScheme("test") // Manual resolver to control the target resolution.
+	target := targetResolver.Scheme() + ":///" + targetTestAddr
+	proxyResolver := setupDNS(t) // Set up a manual DNS resolver to control the proxy address resolution.
+
+	tcc, stateCh, _ := createTestResolverClientConn(t)
+	if _, err := delegatingresolver.New(resolver.Target{URL: *testutils.MustParseURL(target)}, tcc, resolver.BuildOptions{}, targetResolver, false); err != nil {
+		t.Fatalf("Failed to create delegating resolver: %v", err)
+	}
+
+	proxyResolver.UpdateState(resolver.State{
+		Addresses: []resolver.Address{
+			{Addr: resolvedProxyTestAddr1},
+			{Addr: resolvedProxyTestAddr2},
+		},
+		ServiceConfig: &serviceconfig.ParseResult{},
+	})
+
+	select {
+	case <-stateCh:
+		t.Fatalf("Delegating resolver invoked UpdateState before both the proxy and target resolvers had updated their states.")
+	case <-time.After(defaultTestShortTimeout):
+	}
+
+	targetResolver.UpdateState(resolver.State{
+		Addresses: []resolver.Address{
+			{Addr: resolvedTargetTestAddr1},
+			{Addr: resolvedTargetTestAddr2},
+		},
+		ServiceConfig: &serviceconfig.ParseResult{},
+	})
+
+	wantState := resolver.State{
+		Addresses: []resolver.Address{
+			proxyAddressWithTargetAttribute(envProxyAddr, resolvedTargetTestAddr1),
+			proxyAddressWithTargetAttribute(envProxyAddr, resolvedTargetTestAddr2),
+		},
+		ServiceConfig: &serviceconfig.ParseResult{},
+	}
+	var gotState resolver.State
+	select {
+	case gotState = <-stateCh:
+	case <-time.After(defaultTestTimeout):
+		t.Fatal("Timeout when waiting for a state update from the delegating resolver")
+	}
+
+	if diff := cmp.Diff(gotState, wantState); diff != "" {
+		t.Fatalf("Unexpected state from delegating resolver. Diff (-got +want):\n%v\n got state : %s, want state %s\n", diff, pretty.ToJSON(gotState), pretty.ToJSON(wantState))
```

Comment:
We should print only the diff, or the complete output, not both.

Reply:
DOne.

Reply:
`cmp.Diff` output is not always great. Especially it truncates lines when they go over some length. And complete output is also hard sometime when the whole output is huge and you have to squint to figure out where the diff is. So, it does make sense *sometimes* to print both the diff and the complete output. We do it in some of our tests.

I haven't looked at how this error message looks when the test fails, so I can't comment about this particular case. But a blanket statement that we should do only one and not both is not appropriate for all cases.